GVFS/GVFS.Common/NamedPipes/NamedPipeStreamReader.cs

```diff
@@ -0,0 +1,67 @@
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.IO;
+using System.Text;
+
+namespace GVFS.Common.NamedPipes
+{
+    /// <summary>
+    ///   The GVFS NamePipeProtocol:
+    ///       GVFS can send "messages" over the wire. A message is defined as a
+    ///       sequence of UTF8 encoded text whose end is delineated by a TerminatorByte.
+    /// </summary>
+    public class NamedPipeStreamReader
+    {
+        public const int BufferLength = 1024;
+
+        private const byte TerminatorByte = 0x3;
```

Comment:
Have you considered a design where you include the message length as the first part of the message, rather than use a terminator byte? It seems like that could give us better perf when reading messages - especially since there's no need for a `List`, along with dynamic reallocations, and then converting it back to an array afterwards - but I'm not sure how big that perf impact is, necessarily.

Reply:
Yes - I did consider that approach (and in general, I like that approach) - however, it would require more substantive changes. This change focuses on correctness of the underlying protocol to address the associated issue, without regressing performance. This follows the same approach as used by the current `StreamReader` implementation. We have other  more impactful items to address for performance before we need to focus on the wire protocol here (at least, I am not aware of this popping up on any perf investigations).

I do like your suggestion to size the `List` based on the expected message size (or even a focused optimization to avoid allocating a `List` if the entire message is read in the initial read).

Reply:
Why would that be a more substantive change? My naive reading of the code leads me to think that the changes would all be contained in NamedPipeStreamReader/Writer. Are there more complications?

Reply:
There are also native clients that implement this protocol, and they would need to be updated as well.

Reply:
Ah that makes sense, thanks. So I guess the question is: do we think that would be a better solution? And if so, is there still value in building an interim solution here? This is not a pressing need, so I would vote that we build whichever one we think is the better long term fix.